src/main/java/hudson/plugins/ansicolor/ColorConsoleAnnotator.java

```diff
@@ -70,40 +80,67 @@
                         lastPoint = incoming.getCount();
                         int hide = incoming.getCount() - outgoing.getCount() - adjustment;
-                        LOGGER.finest("hiding " + hide + " @" + (outgoing.getCount() + adjustment));
-                        text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
-                        adjustment += hide;
+                        // If openTags is not empty, but there are no escape sequences directly on this line, or if we
+                        // are just closing tags at the end of the line, there is nothing to hide.
+                        if (hide != 0) {
+                            LOGGER.log(Level.FINEST, "hiding {0} @{1}", new Object[] { hide, outgoing.getCount() + adjustment });
+                            text.addMarkup(outgoing.getCount() + adjustment, outgoing.getCount() + adjustment + hide, "<span style=\"display: none\">", "</span>");
+                            adjustment += hide;
+                        }
                     }
                 }
+                public void emitHtmlDirect(String html) {
+                    text.addMarkup(incoming.getCount(), html);
+                }
             }
             EmitterImpl emitter = new EmitterImpl();
-            CountingOutputStream incoming = new CountingOutputStream(new AnsiHtmlOutputStream(outgoing, colorMap, emitter));
+            AnsiHtmlOutputStream ansiOs = new AnsiHtmlOutputStream(outgoing, colorMap, emitter);
+            CountingOutputStream incoming = new CountingOutputStream(ansiOs);
             emitter.incoming = incoming;
             try {
+                for (AnsiAttributeElement element : openTags) {
+                    // We need to reopen tags that were still open at the end of the previous line in the
+                    // so the stream's state is correct in case those tags are closed mid-line.
+                    ansiOs.openTag(element);
+                }
                 byte[] data = s.getBytes(charset);
                 for (int i = 0; i < data.length; i++) {
                     // Do not use write(byte[]) as offsets in incoming would not be accurate.
                     incoming.write(data[i]);
                 }
+                nextOpenTags = ansiOs.getOpenTags();
+                // Close any tags that are still open at the end of the line.
+                ansiOs.closeOpenTags(null);
             } catch (IOException x) {
                 LOGGER.log(Level.WARNING, null, x);
             }
             LOGGER.finer(() -> "\"" + StringEscapeUtils.escapeJava(s) + "\" → \"" + StringEscapeUtils.escapeJava(text.toString(true)) + "\"");
         }
-        return this;
+        return openTags == nextOpenTags
+                ? this
+                : new ColorConsoleAnnotator(colorMapName, charset, nextOpenTags);
+    }
+
+    private Object readResolve() {
```

Comment:
The `ConsoleAnnotator` Javadocs mention serialization, so I added this, although it's not clear to me if serialized `ConsoleAnnotator` objects would survive a restart (in which case this method is needed) or only exist in memory or some other transient form (in which case this method can be deleted).

Reply:
Only in a browser, so I think this could be dropped.

Reply:
Removed in 31bb937.